### ISSUE TITLE
[solved]session->Create(graph_def) No OpKernel was registered to support Op 'RandomUniform'

### ISSUE THREAD

**Issue by eigen2017** [tensorflow/tensorflow#15538]
issue: crashes when loading pb file in C++ program:

>     Session * session;
>     GraphDef graph_def;
>     SessionOptions opts;
>     TF_CHECK_OK(ReadBinaryProto(Env::Default(), heartPrintPbPathFile, &graph_def));
>     TF_CHECK_OK(NewSession(opts, &session));
>     TF_CHECK_OK(session->Create(graph_def));
> 

os: linux
train: python
inference: C++ interface

runtime error on inference part:

> Non-OK-status: session->Create(graph_def) status: Invalid argument: No OpKernel was registered to support Op 'RandomUniform' with these attrs.  Registered devices: [CPU], Registered kernels:
>   <no registered kernels>
> 
> 	 [[Node: rnn_net/rnn/dropout_63/random_uniform/RandomUniform = RandomUniform[T=DT_INT32, dtype=DT_FLOAT, seed=0, seed2=0](rnn_net/rnn/dropout_63/Shape)]]

python train code:

>         with tf.variable_scope("rnn_net") as scope:
>             cell = []
>             for i in range(num_layers):
>                 cell.append(tf.nn.rnn_cell.LSTMCell(hidden_size, state_is_tuple=True))
> 
>             cell = tf.nn.rnn_cell.MultiRNNCell(cell)
> 
>             cell = tf.nn.rnn_cell.DropoutWrapper(cell, output_keep_prob = self.keep_prob)
> 
>             initial_state = cell.zero_state(batch_size, tf.float32)
> 
>             input_list = tf.unstack(conv_output, axis=1)
> 
>             rnn_output, _ = tf.nn.static_rnn(cell, input_list, dtype=tf.float32)
>             self.rnn_output = rnn_output[-1]
>             print "rnn output shape: "
>             print self.rnn_output.get_shape()

hint one : i found that if i delete this line:
cell = tf.nn.rnn_cell.DropoutWrapper(cell, output_keep_prob = self.keep_prob)
then the trained pb file can be loaded normally.

hint two: beside rnn part, i also have cnn part in the network, and the dropout in cnn works just fine:

>         with tf.variable_scope("conv_net") as scope:
>             filters = [15, 11, 7, 5]
>             kernel_size = [64, 64, 32, 32]
>             input_layer = tf.reshape(self.x, [-1, seq_len, 1])
>             conv1_1 = tf.layers.conv1d(inputs=input_layer, filters=filters[0], kernel_size=kernel_size[0], padding="same", activation=tf.nn.tanh, name='conv1_1')
>             conv1_2 = tf.layers.conv1d(inputs=conv1_1, filters=filters[1], kernel_size=kernel_size[1], padding="same", activation=tf.nn.tanh)
>             pool1 = tf.layers.max_pooling1d(inputs=conv1_2, pool_size=4, strides=4)
> 
>             bn1 = batch_norm(pool1, self.is_train, scope='bn1')
>             dropout1 = tf.layers.dropout(inputs=bn1, rate=(1 - self.keep_prob))
> 
>             conv2_1 = tf.layers.conv1d(inputs=dropout1, filters=filters[2], kernel_size=kernel_size[2], padding="same", activation=tf.nn.tanh, name='conv2_1')
>             conv2_2 = tf.layers.conv1d(inputs=conv2_1, filters=filters[3], kernel_size=kernel_size[3], padding="same", activation=tf.nn.tanh)
>             pool2 = tf.layers.max_pooling1d(inputs=conv2_2, pool_size=4, strides=4)
> 
>             bn2 = batch_norm(pool2, self.is_train, scope='bn2')
>             dropout2 = tf.layers.dropout(inputs=bn2, rate=(1 - self.keep_prob))
> 
>             conv3_1 = tf.layers.conv1d(inputs=dropout2, filters=filters[2], kernel_size=kernel_size[2], padding="same", activation=tf.nn.tanh)
>             conv3_2 = tf.layers.conv1d(inputs=conv3_1, filters=filters[3], kernel_size=kernel_size[3], padding="same", activation=tf.nn.tanh)
>             pool3 = tf.layers.max_pooling1d(inputs=conv2_2, pool_size=2, strides=2)
> 
>             conv_output = pool3

**Comment by eigen2017** [comment]
btw: i've been always using this way to build tensor lib :
https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/makefile

**Comment by eigen2017** [comment]
i solved this issue by 
https://dato.ml/drop-dropout-from-frozen-model/

**Comment by eigen2017** [comment]
pb file before change:

> 2338 rnn_net/rnn/rnn/multi_rnn_cell/cell_1/cell_1/lstm_cell/lstm_cell_63/mul_2 Mul
> └─── 0 ─ rnn_net/rnn/rnn/multi_rnn_cell/cell_1/cell_1/lstm_cell/lstm_cell_63/Sigmoid_2
> └─── 1 ─ rnn_net/rnn/rnn/multi_rnn_cell/cell_1/cell_1/lstm_cell/lstm_cell_63/Tanh_1
> 2339 rnn_net/rnn/dropout_63/Shape Shape
> └─── 0 ─ rnn_net/rnn/rnn/multi_rnn_cell/cell_1/cell_1/lstm_cell/lstm_cell_63/mul_2
> 2340 rnn_net/rnn/dropout_63/random_uniform/min Const
> 2341 rnn_net/rnn/dropout_63/random_uniform/max Const
> 2342 rnn_net/rnn/dropout_63/random_uniform/RandomUniform RandomUniform
> └─── 0 ─ rnn_net/rnn/dropout_63/Shape
> 2343 rnn_net/rnn/dropout_63/random_uniform/sub Sub
> └─── 0 ─ rnn_net/rnn/dropout_63/random_uniform/max
> └─── 1 ─ rnn_net/rnn/dropout_63/random_uniform/min
> 2344 rnn_net/rnn/dropout_63/random_uniform/mul Mul
> └─── 0 ─ rnn_net/rnn/dropout_63/random_uniform/RandomUniform
> └─── 1 ─ rnn_net/rnn/dropout_63/random_uniform/sub
> 2345 rnn_net/rnn/dropout_63/random_uniform Add
> └─── 0 ─ rnn_net/rnn/dropout_63/random_uniform/mul
> └─── 1 ─ rnn_net/rnn/dropout_63/random_uniform/min
> 2346 rnn_net/rnn/dropout_63/add Add
> └─── 0 ─ dropout
> └─── 1 ─ rnn_net/rnn/dropout_63/random_uniform
> 2347 rnn_net/rnn/dropout_63/Floor Floor
> └─── 0 ─ rnn_net/rnn/dropout_63/add
> 2348 rnn_net/rnn/dropout_63/div RealDiv
> └─── 0 ─ rnn_net/rnn/rnn/multi_rnn_cell/cell_1/cell_1/lstm_cell/lstm_cell_63/mul_2
> └─── 1 ─ dropout
> 2349 rnn_net/rnn/dropout_63/mul Mul
> └─── 0 ─ rnn_net/rnn/dropout_63/div
> └─── 1 ─ rnn_net/rnn/dropout_63/Floor
> 2350 w1 Const
> 2351 w1/read Identity
> └─── 0 ─ w1
> 2352 b1 Const
> 2353 b1/read Identity
> └─── 0 ─ b1
> 2354 softmax/xw_plus_b/MatMul MatMul
> └─── 0 ─ rnn_net/rnn/dropout_63/mul
> └─── 1 ─ w1/read

**Comment by eigen2017** [comment]
pb file after change:

> 2338 rnn_net/rnn/rnn/multi_rnn_cell/cell_1/cell_1/lstm_cell/lstm_cell_63/mul_2 Mul
> └─── 0 ─ rnn_net/rnn/rnn/multi_rnn_cell/cell_1/cell_1/lstm_cell/lstm_cell_63/Sigmoid_2
> └─── 1 ─ rnn_net/rnn/rnn/multi_rnn_cell/cell_1/cell_1/lstm_cell/lstm_cell_63/Tanh_1
> 2339 w1 Const
> 2340 w1/read Identity
> └─── 0 ─ w1
> 2341 b1 Const
> 2342 b1/read Identity
> └─── 0 ─ b1
> 2343 softmax/xw_plus_b/MatMul MatMul
> └─── 0 ─ rnn_net/rnn/rnn/multi_rnn_cell/cell_1/cell_1/lstm_cell/lstm_cell_63/mul_2
> └─── 1 ─ w1/read

**Comment by michaelisard** [comment]
Closing since this seems resolved: please reopen if I misunderstood!

**Comment by eigen2017** [comment]
yes, it's resolved. @michaelisard
